### PR TITLE
Paranoid fix for removing parts from ZooKeeper

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Slightly related to https://github.com/ClickHouse/ClickHouse/issues/50341

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit f62faaedc3c837b009479cb971e868526a6464fa with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50448/f62faaedc3c837b009479cb971e868526a6464fa/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by alexey-milovidov — APPROVED** [review]

File: src/Storages/MergeTree/ReplicatedMergeTreeSink.h [hunk 1]
```diff
@@ -87,7 +87,7 @@ class ReplicatedMergeTreeSinkImpl : public SinkToStorage
     size_t checkQuorumPrecondition(const ZooKeeperWithFaultInjectionPtr & zookeeper);
 
     /// Rename temporary part and commit to ZooKeeper.
-    std::vector<String> commitPart(
+    std::pair<std::vector<String>, bool> commitPart(
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/ReplicatedMergeTreeSink.h:90]: I'd prefer to have a better comment here.

**Review by alexey-milovidov — APPROVED** [review]
LGTM.

But what is the point? Does it mean that if an insertion was retried after ZooKeeper failure and then deduplicated, an orphan part may remain, and require additional effort on deletion?

**Comment by tavplubix** [comment]
> But what is the point? Does it mean that if an insertion was retried after ZooKeeper failure and then deduplicated, an orphan part may remain, and require additional effort on deletion?

I don't know but looks like in some rare cases some removed parts may remain in ZooKeeper leading to false-positive "part is lost forever" messages. It's interesting that it usually happens to zero-level parts.

**Pull request merged**